### PR TITLE
Improve OpenGL Rectangle SetInstances Performance

### PR DESCRIPTION
**Pull request by ccahiggins**
Change to use glBufferSubData instead of glBufferData

An initial buffer size is set with glBufferData.
glBufferSubData is then used to copy to the existing buffer, rather than creating a new one for each frame.

We keep track of the buffer size, and if instances size increases above the buffer size, a new buffer is created with glBufferData

### CODE REVIEW

**Comment by ccahiggins** [comment]
On my computer, max zoom on a Crazy Castle save, DrawRectShader::SetInstances takes about 8-10ms to complete. This drops is down to 2.5-3ms. Frame rate at max zoom, with multithreading, is now just over 50FPS - 40 previously.

Prison Island is even slower, 18FPS before, and about 3FPS higher with this.

File: src/openrct2-ui/drawing/engines/opengl/DrawRectShader.cpp [hunk 1]
```diff
@@ -11,8 +11,12 @@
 
 #    include "DrawRectShader.h"
 
+#    define INITIAL_INSTANCES_BUFFER_SIZE 32768
```
Comment (duncanspumpkin) [inline src/openrct2-ui/drawing/engines/opengl/DrawRectShader.cpp:14]: Don't use #defines use constexpr

**Comment by duncanspumpkin** [comment]
Are there any links to read about why this works. Are there any negatives with this? Is it cross platform

**Comment by ccahiggins** [comment]
https://registry.khronos.org/OpenGL-Refpages/gl4/html/glBufferSubData.xhtml
https://handmade.network/forums/t/1522-glbufferdata_causing_time_spikes

Both suggest using glBufferSubData instead of destroying/recreating the buffer each frame with glBufferData.

Could behave differently with different graphics drivers, I've only tested this in Windows on my 3080. Definitely would be good to test this on other platforms.

glBufferSubData blocks the GPU while it's called, so if the GPU hasn't finished drawing from the buffer when it's called, it can stall. I'm not sure if this would be an issue with OpenRCT2 as the buffer is only updated at the end of the frame anyway.

**Comment by ccahiggins** [comment]
I just got around to testing this on a laptop with Intel graphics, and the framerate was actually a bit lower.

**Comment by janisozaur** [comment]
I'd like to test it as well, can you please give me time until weekend?

**Review by Gymnasiast — COMMENTED** [review]
Seems to work correctly on Ubuntu 20.04 using Intel HD Graphics 530. Haven’t noticed any performance improvement, though. Would like someone to verify that before merging this.

**Comment by ccahiggins** [comment]
Experiencied some weird issues tonight when switching between Software and OpenGL. Only when the window was maximised to 1440p, but this doesn't look fully stable.

Glitch

**Comment by 733737** [comment]
could be a fluke

**Comment by ccahiggins** [comment]
Not a fluke, I could consistently recreate it. Should be fixed now though

**Comment by janisozaur** [comment]
Sorry, haven't tested it yet, will do that today

File: src/openrct2-ui/drawing/engines/opengl/DrawRectShader.cpp [hunk 2]
```diff
@@ -27,11 +27,15 @@ constexpr VDStruct VertexData[4] = {
     { 0.0f, 0.0f, 0.0f, 0.0f, 1.0f, 0.0f, 0.0f, 1.0f, 1.0f, 1.0f },
 };
 
+constexpr const size_t InitialInstancesBufferSize = 32768;
+
 DrawRectShader::DrawRectShader()
     : OpenGLShaderProgram("drawrect")
```
Comment (janisozaur) [inline src/openrct2-ui/drawing/engines/opengl/DrawRectShader.cpp:33]: Can you move assignment to initialization list?

**Review by janisozaur — APPROVED** [review]
Just a minor change request, but tested it and looks good overall.

**Comment by ccahiggins** [comment]
> Just a minor change request, but tested it and looks good overall.

Did you notice any difference in performance? Don't want to negatively affect lower end systems.

I tested on my 3080 and, most notably at max zoom it made a difference. Seemed ever so slightly slower on my work laptop. I think I still have my 2070 somewhere which I can also test with, and maybe - assuming I didn't chuck it, or the old motherboard out - a Radeon HD 6750 (the old 6750 not the new one).

File: src/openrct2-ui/drawing/engines/opengl/DrawRectShader.cpp (hunk 2 above)
Comment (ccahiggins) [reply src/openrct2-ui/drawing/engines/opengl/DrawRectShader.cpp:33]: Yeah, that makes sense, will do

**Comment by karst** [comment]
Out of the testing I have done, comparing it to develop, I have about an 18.5% increase in average performance using this build.

**Review by Gymnasiast — APPROVED** [review]

**Pull request merged**